pallets/staking/src/lib.rs

```diff
@@ -0,0 +1,926 @@
+		#[pallet::constant]
+		type MaxVotes: Get<u32>;
+
+		/// Democracy referendum state.
+		type ReferendumInfo: DemocracyReferendum;
+
+		/// Amount of action points user will receive for each voted HDX.
+		type ActionMultiplier: GetByKey<Action, u32>;
+
+		/// Provides information about amount of vested tokens.
+		type Vesting: VestingDetails<Self::AccountId, Balance>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// Unit we are distributing action points for.
+		/// e.g if RewardedVoteUnit is 1HDX user will receive `x` action points per each voted 1 HDX.
+		#[pallet::constant]
+		type RewardedVoteUnit: Get<Balance>;
+	}
+
+	#[pallet::storage]
+	/// Global staking state.
+	#[pallet::getter(fn staking)]
+	pub(super) type Staking<T: Config> = StorageValue<_, StakingData, ValueQuery>;
+
+	#[pallet::storage]
+	/// User's position state.
+	#[pallet::getter(fn positions)]
+	pub(super) type Positions<T: Config> = StorageMap<_, Blake2_128Concat, T::PositionItemId, Position<T::BlockNumber>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_position_id)]
+	/// Position ids sequencer.
+	pub(super) type NextPositionId<T: Config> = StorageValue<_, T::PositionItemId, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn position_votes)]
+	/// List of position votes.
+	pub(super) type PositionVotes<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::PositionItemId, Voting<T::MaxVotes>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// New staking position was created and NFT was minted.
+		PositionCreated {
+			who: T::AccountId,
+			position_id: T::PositionItemId,
+			stake: Balance,
+		},
+
+		/// Staked amount for existing position was increased.
+		StakeAdded {
+			who: T::AccountId,
+			position_id: T::PositionItemId,
+			stake: Balance,
+			total_stake: Balance,
+			locked_rewards: Balance,
+			slashed_points: Point,
+		},
+
+		/// Rewards was claimed.
+		RewardsClaimed {
+			who: T::AccountId,
+			position_id: T::PositionItemId,
+			paid_rewards: Balance,
+			unlocked_rewards: Balance,
+			slashed_points: Point,
+			slashed_unpaid_rewards: Balance,
+		},
+
+		/// Staked amount was withdrawn and NFT was burned.
+		Unstaked {
+			who: T::AccountId,
+			position_id: T::PositionItemId,
+			unlocked_stake: Balance,
+			rewards: Balance,
+			unlocked_rewards: Balance,
+		},
+
+		/// Staking was initialized.
+		StakingInitialized { non_dustable_balance: Balance },
+
+		/// Staking's `accumulated_reward_per_stake` was updated.
+		AccumulatedRpsUpdated {
+			accumulated_rps: FixedU128,
+			total_stake: Balance,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Balance is too low.
+		InsufficientBalance,
+
+		/// Staked amount is too low.
+		InsufficientStake,
+
+		/// Staking position has not been found.
+		PositionNotFound,
+
+		/// Max amount of votes was reached for staking position.
+		MaxVotesReached,
```

Comment:
Are we using this anywhere? Didn't find a reference.

Reply:
it's used in democracy integration e.g here: https://github.com/galacticcouncil/HydraDX-node/blob/a1fa93ac93da7173893e0ee32bb0c8d6396f6dbd/pallets/staking/src/integrations/democracy.rs#L68